{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3862f5b4",
   "metadata": {},
   "source": [
    "## Exporting scRNAseq data using Python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "08538082",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_376608/3713809626.py:11: ImplicitModificationWarning: Trying to modify attribute `.var` of view, initializing view as actual.\n",
      "  adata_ref.var['mt'] = adata_ref.var_names.str.upper().str.startswith('MT-')\n"
     ]
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "\n",
    "# read h5ad files.\n",
    "adata_ref = sc.read_h5ad(\"/disk2/user/hilpek/project_root/scRNAseq/input/Wu_all_NT.h5ad\")\n",
    "\n",
    "# Removed HER samples.\n",
    "adata_ref = adata_ref[(adata_ref.obs['IHC_subtype'] == 'ER+'), :]\n",
    "\n",
    "# mt and ribo columns have been added to adata.var\n",
    "adata_ref.var['mt'] = adata_ref.var_names.str.upper().str.startswith('MT-')\n",
    "adata_ref.var['ribo'] = adata_ref.var_names.str.upper().str.startswith(('RPL', 'RPS'))\n",
    "adata_ref.var_names_make_unique()\n",
    "\n",
    "# Get ham counts matrix as a dataframe with gene and cell types.\n",
    "import numpy as np\n",
    "counts = pd.DataFrame(\n",
    "    adata_ref.X.toarray() if hasattr(adata_ref.X, \"toarray\") else np.array(adata_ref.X),\n",
    "    index=adata_ref.obs_names,\n",
    "    columns=adata_ref.var_names\n",
    ")\n",
    "counts.T.to_csv(\"/disk2/user/hilpek/project_root/scRNAseq/input2/dge.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e0a6a421",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 33790 × 29733\n",
       "    obs: 'Study_id', 'Celltype_subset', 'bc_subtype', 'donor_age', 'treatment_status', 'grade', 'IHC_subtype', 'Dataset'\n",
       "    var: 'ENSEMBL', 'mt', 'ribo'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d37c6418",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Study_id</th>\n",
       "      <th>Celltype_subset</th>\n",
       "      <th>bc_subtype</th>\n",
       "      <th>donor_age</th>\n",
       "      <th>treatment_status</th>\n",
       "      <th>grade</th>\n",
       "      <th>IHC_subtype</th>\n",
       "      <th>Dataset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CID3941_AAAGTAGTCCTTGCCA</th>\n",
       "      <td>CID3941</td>\n",
       "      <td>Endothelial ACKR1</td>\n",
       "      <td>IDC</td>\n",
       "      <td>50.0</td>\n",
       "      <td>Naive</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ER+</td>\n",
       "      <td>CID3941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CID3941_ACAGCCGAGCTCTCGG</th>\n",
       "      <td>CID3941</td>\n",
       "      <td>Endothelial ACKR1</td>\n",
       "      <td>IDC</td>\n",
       "      <td>50.0</td>\n",
       "      <td>Naive</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ER+</td>\n",
       "      <td>CID3941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CID3941_ACGCAGCAGGACCACA</th>\n",
       "      <td>CID3941</td>\n",
       "      <td>Endothelial ACKR1</td>\n",
       "      <td>IDC</td>\n",
       "      <td>50.0</td>\n",
       "      <td>Naive</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ER+</td>\n",
       "      <td>CID3941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CID3941_ACGCAGCCAGCTGTAT</th>\n",
       "      <td>CID3941</td>\n",
       "      <td>Endothelial ACKR1</td>\n",
       "      <td>IDC</td>\n",
       "      <td>50.0</td>\n",
       "      <td>Naive</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ER+</td>\n",
       "      <td>CID3941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CID3941_ACTTTCACAGCCTGTG</th>\n",
       "      <td>CID3941</td>\n",
       "      <td>Endothelial ACKR1</td>\n",
       "      <td>IDC</td>\n",
       "      <td>50.0</td>\n",
       "      <td>Naive</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ER+</td>\n",
       "      <td>CID3941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CID4535_TAGTGGTCATTGGCGC</th>\n",
       "      <td>CID4535</td>\n",
       "      <td>Mature Luminal</td>\n",
       "      <td>ILC</td>\n",
       "      <td>47.0</td>\n",
       "      <td>Naive</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ER+</td>\n",
       "      <td>CID4535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CID4535_TCTATTGAGCTAGGCA</th>\n",
       "      <td>CID4535</td>\n",
       "      <td>Mature Luminal</td>\n",
       "      <td>ILC</td>\n",
       "      <td>47.0</td>\n",
       "      <td>Naive</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ER+</td>\n",
       "      <td>CID4535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CID4535_TGCTACCAGGTGTGGT</th>\n",
       "      <td>CID4535</td>\n",
       "      <td>Mature Luminal</td>\n",
       "      <td>ILC</td>\n",
       "      <td>47.0</td>\n",
       "      <td>Naive</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ER+</td>\n",
       "      <td>CID4535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CID4535_TGGCCAGAGCCCAACC</th>\n",
       "      <td>CID4535</td>\n",
       "      <td>Mature Luminal</td>\n",
       "      <td>ILC</td>\n",
       "      <td>47.0</td>\n",
       "      <td>Naive</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ER+</td>\n",
       "      <td>CID4535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CID4535_TTGGAACTCGGTCCGA</th>\n",
       "      <td>CID4535</td>\n",
       "      <td>Mature Luminal</td>\n",
       "      <td>ILC</td>\n",
       "      <td>47.0</td>\n",
       "      <td>Naive</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ER+</td>\n",
       "      <td>CID4535</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33790 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         Study_id    Celltype_subset bc_subtype  donor_age  \\\n",
       "CID3941_AAAGTAGTCCTTGCCA  CID3941  Endothelial ACKR1        IDC       50.0   \n",
       "CID3941_ACAGCCGAGCTCTCGG  CID3941  Endothelial ACKR1        IDC       50.0   \n",
       "CID3941_ACGCAGCAGGACCACA  CID3941  Endothelial ACKR1        IDC       50.0   \n",
       "CID3941_ACGCAGCCAGCTGTAT  CID3941  Endothelial ACKR1        IDC       50.0   \n",
       "CID3941_ACTTTCACAGCCTGTG  CID3941  Endothelial ACKR1        IDC       50.0   \n",
       "...                           ...                ...        ...        ...   \n",
       "CID4535_TAGTGGTCATTGGCGC  CID4535     Mature Luminal        ILC       47.0   \n",
       "CID4535_TCTATTGAGCTAGGCA  CID4535     Mature Luminal        ILC       47.0   \n",
       "CID4535_TGCTACCAGGTGTGGT  CID4535     Mature Luminal        ILC       47.0   \n",
       "CID4535_TGGCCAGAGCCCAACC  CID4535     Mature Luminal        ILC       47.0   \n",
       "CID4535_TTGGAACTCGGTCCGA  CID4535     Mature Luminal        ILC       47.0   \n",
       "\n",
       "                         treatment_status  grade IHC_subtype  Dataset  \n",
       "CID3941_AAAGTAGTCCTTGCCA            Naive    2.0         ER+  CID3941  \n",
       "CID3941_ACAGCCGAGCTCTCGG            Naive    2.0         ER+  CID3941  \n",
       "CID3941_ACGCAGCAGGACCACA            Naive    2.0         ER+  CID3941  \n",
       "CID3941_ACGCAGCCAGCTGTAT            Naive    2.0         ER+  CID3941  \n",
       "CID3941_ACTTTCACAGCCTGTG            Naive    2.0         ER+  CID3941  \n",
       "...                                   ...    ...         ...      ...  \n",
       "CID4535_TAGTGGTCATTGGCGC            Naive    2.0         ER+  CID4535  \n",
       "CID4535_TCTATTGAGCTAGGCA            Naive    2.0         ER+  CID4535  \n",
       "CID4535_TGCTACCAGGTGTGGT            Naive    2.0         ER+  CID4535  \n",
       "CID4535_TGGCCAGAGCCCAACC            Naive    2.0         ER+  CID4535  \n",
       "CID4535_TTGGAACTCGGTCCGA            Naive    2.0         ER+  CID4535  \n",
       "\n",
       "[33790 rows x 8 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_ref.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb945447",
   "metadata": {},
   "source": [
    "#### Get major cell types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0108b06a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Endothelial ACKR1': 'Endothelial',\n",
       " 'Endothelial RGS5': 'Endothelial',\n",
       " 'Endothelial CXCL12': 'Endothelial',\n",
       " 'CAFs MSC iCAF-like s1': 'CAFs',\n",
       " 'CAFs MSC iCAF-like s2': 'CAFs',\n",
       " 'CAFs Transitioning s3': 'CAFs',\n",
       " 'CAFs myCAF like s4': 'CAFs',\n",
       " 'CAFs myCAF like s5': 'CAFs',\n",
       " 'PVL Differentiated s3': 'PVL',\n",
       " 'PVL_Immature s2': 'PVL',\n",
       " 'PVL Immature s1': 'PVL',\n",
       " 'Endothelial Lymphatic LYVE1': 'Endothelial',\n",
       " 'B cells Memory': 'B-cells',\n",
       " 'B cells Naive': 'B-cells',\n",
       " 'T_cells_c4_CD8+_ZFP36': 'T-cells',\n",
       " 'T_cells_c6_IFIT1': 'T-cells',\n",
       " 'T_cells_c5_CD8+_GZMK': 'T-cells',\n",
       " 'T_cells_c7_CD8+_IFNG': 'T-cells',\n",
       " 'T_cells_c8_CD8+_LAG3': 'T-cells',\n",
       " 'T_cells_c0_CD4+_CCR7': 'T-cells',\n",
       " 'T_cells_c1_CD4+_IL7R': 'T-cells',\n",
       " 'T_cells_c2_CD4+_T-regs_FOXP3': 'T-cells',\n",
       " 'T_cells_c3_CD4+_Tfh_CXCL13': 'T-cells',\n",
       " 'T_cells_c9_NK_cells_AREG': 'T-cells',\n",
       " 'T_cells_c11_MKI67': 'T-cells',\n",
       " 'T_cells_c10_NKT_cells_FCGR3A': 'T-cells',\n",
       " 'Myeloid_c10_Macrophage_1_EGR1': 'Myeloid',\n",
       " 'Myeloid_c12_Monocyte_1_IL1B': 'Myeloid',\n",
       " 'Myeloid_c2_LAM2_APOE': 'Myeloid',\n",
       " 'Myeloid_c1_LAM1_FABP5': 'Myeloid',\n",
       " 'Myeloid_c8_Monocyte_2_S100A9': 'Myeloid',\n",
       " 'Myeloid_c7_Monocyte_3_FCGR3A': 'Myeloid',\n",
       " 'Myeloid_c9_Macrophage_2_CXCL10': 'Myeloid',\n",
       " 'Cycling_Myeloid': 'Myeloid',\n",
       " 'Myeloid_c11_cDC2_CD1C': 'Myeloid',\n",
       " 'Myeloid_c4_DCs_pDC_IRF7': 'Myeloid',\n",
       " 'Myeloid_c3_cDC1_CLEC9A': 'Myeloid',\n",
       " 'Myeloid_c0_DC_LAMP3': 'Myeloid',\n",
       " 'Myoepithelial': 'Normal Epithelial',\n",
       " 'Luminal Progenitors': 'Normal Epithelial',\n",
       " 'Mature Luminal': 'Normal Epithelial',\n",
       " 'Plasmablasts': 'Plasmablasts',\n",
       " 'Cancer Cycling': 'Cancer Epithelial',\n",
       " 'Cancer Her2 SC': 'Cancer Epithelial',\n",
       " 'Cancer LumB SC': 'Cancer Epithelial',\n",
       " 'Cancer Basal SC': 'Cancer Epithelial',\n",
       " 'Cycling PVL': 'PVL',\n",
       " 'Cancer LumA SC': 'Cancer Epithelial',\n",
       " 'Myeloid_c5_Macrophage_3_SIGLEC1': 'Myeloid'}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load metadata\n",
    "metadata = pd.read_csv(\"/disk2/user/hilpek/project_root/scRNAseq/input/metadata.csv\", sep=\",\")\n",
    "metadata.set_index('Unnamed: 0', inplace=True)\n",
    "metadata\n",
    "\n",
    "# get corresponding major cell types for each subset of cell type\n",
    "subset_to_major_dict = {}\n",
    "for index, row in metadata.iterrows():\n",
    "    subset = row[\"celltype_subset\"]\n",
    "    major = row[\"celltype_major\"]\n",
    "    if subset not in subset_to_major_dict.keys():\n",
    "        subset_to_major_dict[subset] = major\n",
    "\n",
    "subset_to_major_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a5eb3228",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Study_id</th>\n",
       "      <th>Celltype_subset</th>\n",
       "      <th>bc_subtype</th>\n",
       "      <th>donor_age</th>\n",
       "      <th>treatment_status</th>\n",
       "      <th>grade</th>\n",
       "      <th>IHC_subtype</th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Celltype_major</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CID3941_AAAGTAGTCCTTGCCA</th>\n",
       "      <td>CID3941</td>\n",
       "      <td>Endothelial ACKR1</td>\n",
       "      <td>IDC</td>\n",
       "      <td>50.0</td>\n",
       "      <td>Naive</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ER+</td>\n",
       "      <td>CID3941</td>\n",
       "      <td>Endothelial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CID3941_ACAGCCGAGCTCTCGG</th>\n",
       "      <td>CID3941</td>\n",
       "      <td>Endothelial ACKR1</td>\n",
       "      <td>IDC</td>\n",
       "      <td>50.0</td>\n",
       "      <td>Naive</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ER+</td>\n",
       "      <td>CID3941</td>\n",
       "      <td>Endothelial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CID3941_ACGCAGCAGGACCACA</th>\n",
       "      <td>CID3941</td>\n",
       "      <td>Endothelial ACKR1</td>\n",
       "      <td>IDC</td>\n",
       "      <td>50.0</td>\n",
       "      <td>Naive</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ER+</td>\n",
       "      <td>CID3941</td>\n",
       "      <td>Endothelial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CID3941_ACGCAGCCAGCTGTAT</th>\n",
       "      <td>CID3941</td>\n",
       "      <td>Endothelial ACKR1</td>\n",
       "      <td>IDC</td>\n",
       "      <td>50.0</td>\n",
       "      <td>Naive</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ER+</td>\n",
       "      <td>CID3941</td>\n",
       "      <td>Endothelial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CID3941_ACTTTCACAGCCTGTG</th>\n",
       "      <td>CID3941</td>\n",
       "      <td>Endothelial ACKR1</td>\n",
       "      <td>IDC</td>\n",
       "      <td>50.0</td>\n",
       "      <td>Naive</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ER+</td>\n",
       "      <td>CID3941</td>\n",
       "      <td>Endothelial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CID4535_TAGTGGTCATTGGCGC</th>\n",
       "      <td>CID4535</td>\n",
       "      <td>Mature Luminal</td>\n",
       "      <td>ILC</td>\n",
       "      <td>47.0</td>\n",
       "      <td>Naive</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ER+</td>\n",
       "      <td>CID4535</td>\n",
       "      <td>Normal Epithelial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CID4535_TCTATTGAGCTAGGCA</th>\n",
       "      <td>CID4535</td>\n",
       "      <td>Mature Luminal</td>\n",
       "      <td>ILC</td>\n",
       "      <td>47.0</td>\n",
       "      <td>Naive</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ER+</td>\n",
       "      <td>CID4535</td>\n",
       "      <td>Normal Epithelial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CID4535_TGCTACCAGGTGTGGT</th>\n",
       "      <td>CID4535</td>\n",
       "      <td>Mature Luminal</td>\n",
       "      <td>ILC</td>\n",
       "      <td>47.0</td>\n",
       "      <td>Naive</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ER+</td>\n",
       "      <td>CID4535</td>\n",
       "      <td>Normal Epithelial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CID4535_TGGCCAGAGCCCAACC</th>\n",
       "      <td>CID4535</td>\n",
       "      <td>Mature Luminal</td>\n",
       "      <td>ILC</td>\n",
       "      <td>47.0</td>\n",
       "      <td>Naive</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ER+</td>\n",
       "      <td>CID4535</td>\n",
       "      <td>Normal Epithelial</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CID4535_TTGGAACTCGGTCCGA</th>\n",
       "      <td>CID4535</td>\n",
       "      <td>Mature Luminal</td>\n",
       "      <td>ILC</td>\n",
       "      <td>47.0</td>\n",
       "      <td>Naive</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ER+</td>\n",
       "      <td>CID4535</td>\n",
       "      <td>Normal Epithelial</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33790 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         Study_id    Celltype_subset bc_subtype  donor_age  \\\n",
       "CID3941_AAAGTAGTCCTTGCCA  CID3941  Endothelial ACKR1        IDC       50.0   \n",
       "CID3941_ACAGCCGAGCTCTCGG  CID3941  Endothelial ACKR1        IDC       50.0   \n",
       "CID3941_ACGCAGCAGGACCACA  CID3941  Endothelial ACKR1        IDC       50.0   \n",
       "CID3941_ACGCAGCCAGCTGTAT  CID3941  Endothelial ACKR1        IDC       50.0   \n",
       "CID3941_ACTTTCACAGCCTGTG  CID3941  Endothelial ACKR1        IDC       50.0   \n",
       "...                           ...                ...        ...        ...   \n",
       "CID4535_TAGTGGTCATTGGCGC  CID4535     Mature Luminal        ILC       47.0   \n",
       "CID4535_TCTATTGAGCTAGGCA  CID4535     Mature Luminal        ILC       47.0   \n",
       "CID4535_TGCTACCAGGTGTGGT  CID4535     Mature Luminal        ILC       47.0   \n",
       "CID4535_TGGCCAGAGCCCAACC  CID4535     Mature Luminal        ILC       47.0   \n",
       "CID4535_TTGGAACTCGGTCCGA  CID4535     Mature Luminal        ILC       47.0   \n",
       "\n",
       "                         treatment_status  grade IHC_subtype  Dataset  \\\n",
       "CID3941_AAAGTAGTCCTTGCCA            Naive    2.0         ER+  CID3941   \n",
       "CID3941_ACAGCCGAGCTCTCGG            Naive    2.0         ER+  CID3941   \n",
       "CID3941_ACGCAGCAGGACCACA            Naive    2.0         ER+  CID3941   \n",
       "CID3941_ACGCAGCCAGCTGTAT            Naive    2.0         ER+  CID3941   \n",
       "CID3941_ACTTTCACAGCCTGTG            Naive    2.0         ER+  CID3941   \n",
       "...                                   ...    ...         ...      ...   \n",
       "CID4535_TAGTGGTCATTGGCGC            Naive    2.0         ER+  CID4535   \n",
       "CID4535_TCTATTGAGCTAGGCA            Naive    2.0         ER+  CID4535   \n",
       "CID4535_TGCTACCAGGTGTGGT            Naive    2.0         ER+  CID4535   \n",
       "CID4535_TGGCCAGAGCCCAACC            Naive    2.0         ER+  CID4535   \n",
       "CID4535_TTGGAACTCGGTCCGA            Naive    2.0         ER+  CID4535   \n",
       "\n",
       "                             Celltype_major  \n",
       "CID3941_AAAGTAGTCCTTGCCA        Endothelial  \n",
       "CID3941_ACAGCCGAGCTCTCGG        Endothelial  \n",
       "CID3941_ACGCAGCAGGACCACA        Endothelial  \n",
       "CID3941_ACGCAGCCAGCTGTAT        Endothelial  \n",
       "CID3941_ACTTTCACAGCCTGTG        Endothelial  \n",
       "...                                     ...  \n",
       "CID4535_TAGTGGTCATTGGCGC  Normal Epithelial  \n",
       "CID4535_TCTATTGAGCTAGGCA  Normal Epithelial  \n",
       "CID4535_TGCTACCAGGTGTGGT  Normal Epithelial  \n",
       "CID4535_TGGCCAGAGCCCAACC  Normal Epithelial  \n",
       "CID4535_TTGGAACTCGGTCCGA  Normal Epithelial  \n",
       "\n",
       "[33790 rows x 9 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# save into adata.obs\n",
    "adata_ref.obs['Celltype_major'] = adata_ref.obs['Celltype_subset'].map(subset_to_major_dict)\n",
    "adata_ref.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "acb26195",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 33790 × 29733\n",
       "    obs: 'Study_id', 'Celltype_subset', 'bc_subtype', 'donor_age', 'treatment_status', 'grade', 'IHC_subtype', 'Dataset', 'Celltype_major'\n",
       "    var: 'ENSEMBL', 'mt', 'ribo'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0148d41e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# calculate nUMI\n",
    "nUMI = adata_ref.X.sum(axis=1)\n",
    "if hasattr(nUMI, \"A1\"):\n",
    "    nUMI = nUMI.A1\n",
    "else:\n",
    "    nUMI = np.array(nUMI).flatten()\n",
    "\n",
    "# create metadata\n",
    "meta = pd.DataFrame({\n",
    "    \"barcode\": adata_ref.obs_names.values,                      # cell barcodes\n",
    "    \"cluster\": adata_ref.obs[\"Celltype_subset\"].values,         # cell types\n",
    "    \"nUMI\": nUMI                                                # total UMI\n",
    "})\n",
    "\n",
    "# save file\n",
    "meta.to_csv(\"/disk2/user/hilpek/project_root/scRNAseq/input2/meta_data.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "749d546a",
   "metadata": {},
   "source": [
    "## Import the raw count matrix and spatial coordinates into R"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "35a17236",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/disk2/user/hilpek/miniconda3/envs/cell2_env/lib/python3.9/site-packages/xarray_schema/__init__.py:1: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  from pkg_resources import DistributionNotFound, get_distribution\n",
      "/disk2/user/hilpek/miniconda3/envs/cell2_env/lib/python3.9/site-packages/numba/core/decorators.py:246: RuntimeWarning: nopython is set for njit and is ignored\n",
      "  warnings.warn('nopython is set for njit and is ignored', RuntimeWarning)\n"
     ]
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import squidpy as sq\n",
    "from scipy.sparse import csr_matrix\n",
    "import anndata as ad\n",
    "import os\n",
    "import seaborn as sns\n",
    "from sklearn.metrics import confusion_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3ac9e81e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/disk2/user/hilpek/miniconda3/envs/cell2_env/lib/python3.9/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n",
      "/disk2/user/hilpek/miniconda3/envs/cell2_env/lib/python3.9/site-packages/anndata/_core/anndata.py:1820: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"var\")\n"
     ]
    }
   ],
   "source": [
    "import squidpy as sq\n",
    "adata_vis = sq.read.visium(\"/disk3/cda/SpatialTranscriptomics/processed-data/SpaceRanger-output_v1.0.0-rerun2023/V10F03-034_C/outs\",\n",
    "        counts_file='filtered_feature_bc_matrix.h5', \n",
    "        source_image_path=\"/disk3/cda/SpatialTranscriptomics/raw-data/High-resolution_tissue_images/V10F03-034/210223_BC_S7_V10F03-034_RJ.C1-Spot000001.jpg\"\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "809ed1a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the sample id for each spot\n",
    "adata_vis.obs['sample'] = list(adata_vis.uns['spatial'].keys())[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "7ea236bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add pathologist annotations\n",
    "annoDf = pd.read_csv(\"/disk3/cda/SpatialTranscriptomics/pathologist_annotations/Tissue-annotations_pathologist-2-and-3_CSVs/V10F03-034_C_S7_Wenwen-annotations.csv\")\n",
    "\n",
    "# replace empty annotations with \"Mixed\"\n",
    "annoDf.fillna('Mixed', inplace=True)\n",
    "annoArray = annoDf[\"Wenwen annotations\"].to_numpy()\n",
    "adata_vis.obs[\"Pathologist_annotations\"] = pd.Categorical(annoArray)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "2bef6b1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# add sample name to obs names, important for looking at individual slides later\n",
    "adata_vis.obs[\"sample\"] = [str(i) for i in adata_vis.obs['sample']]\n",
    "adata_vis.obs_names = adata_vis.obs[\"sample\"] + '_' + adata_vis.obs_names\n",
    "adata_vis.obs.index.name = 'spot_id'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ed130a3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_vis.var['SYMBOL'] = adata_vis.var_names\n",
    "adata_vis.var.set_index('gene_ids', drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "91140f62",
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculate QC metrics\n",
    "if not isinstance(adata_vis.X, np.ndarray):\n",
    "    adata_vis.X = adata_vis.X.toarray()\n",
    "    \n",
    "adata_vis.var['mt'] = adata_vis.var[\"SYMBOL\"].str.startswith((\"mt-\", \"MT-\"))\n",
    "adata_vis.var[\"ribo\"] = adata_vis.var[\"SYMBOL\"].str.startswith((\"RPS\", \"RPL\"))\n",
    "\n",
    "sc.pp.calculate_qc_metrics(adata_vis, qc_vars=['mt', 'ribo'], percent_top=None, log1p=False, inplace=True)\n",
    "\n",
    "adata_vis.X = csr_matrix(adata_vis.X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "226caf10",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find and remove MT genes for spatial mapping (keeping their counts in the object)\n",
    "adata_vis.obs['mt_frac'] = adata_vis[:, adata_vis.var['mt'].tolist()].X.sum(1).A.squeeze()/adata_vis.obs['total_counts']\n",
    "adata_vis.obsm['MT'] = adata_vis[:, adata_vis.var['mt'].values].X.toarray()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "f05b6f4f",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_vis = adata_vis[:, ~adata_vis.var['mt'].values]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "3e363719",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_376608/2966163283.py:2: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata_vis.obs['ribo_frac'] = adata_vis[:, adata_vis.var['ribo'].tolist()].X.sum(1).A.squeeze()/adata_vis.obs['total_counts']\n"
     ]
    }
   ],
   "source": [
    "# find and remove RB genes for spatial mapping (keeping their counts in the objm)\n",
    "adata_vis.obs['ribo_frac'] = adata_vis[:, adata_vis.var['ribo'].tolist()].X.sum(1).A.squeeze()/adata_vis.obs['total_counts']\n",
    "adata_vis.obsm['RB'] = adata_vis[:, adata_vis.var['ribo'].values].X.toarray()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "da6ea70f",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_vis = adata_vis[:, ~adata_vis.var['ribo'].values]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "fe2236fc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/disk2/user/hilpek/miniconda3/envs/cell2_env/lib/python3.9/site-packages/scanpy/preprocessing/_simple.py:167: ImplicitModificationWarning: Trying to modify attribute `.obs` of view, initializing view as actual.\n",
      "  adata.obs[\"n_genes\"] = number\n",
      "/disk2/user/hilpek/miniconda3/envs/cell2_env/lib/python3.9/site-packages/scanpy/preprocessing/_simple.py:283: ImplicitModificationWarning: Trying to modify attribute `.var` of view, initializing view as actual.\n",
      "  adata.var[\"n_cells\"] = number\n"
     ]
    }
   ],
   "source": [
    " # choose appropriate thresholds for ST data\n",
    "sc.pp.filter_cells(adata_vis, min_genes = 200)\n",
    "sc.pp.filter_cells(adata_vis, max_counts= 3000)\n",
    "adata_vis = adata_vis[adata_vis.obs[\"pct_counts_mt\"] < 5]\n",
    "\n",
    "# filter genes that are only present in a few spots\n",
    "sc.pp.filter_genes(adata_vis, min_cells=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "315f1a9b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Spot × gene counts\n",
    "pd.DataFrame(\n",
    "    adata_vis.X.toarray() if hasattr(adata_vis.X, \"toarray\") else adata_vis.X.A,\n",
    "    index=adata_vis.obs_names,\n",
    "    columns=adata_vis.var_names\n",
    ").T.to_csv(\"/disk2/user/hilpek/project_root/scRNAseq/input2/MappedDGEForR.csv\")\n",
    "\n",
    "# coordinates\n",
    "coords = adata_vis.obsm['spatial']\n",
    "coords_df = pd.DataFrame(coords, columns=[\"x\", \"y\"], index=adata_vis.obs_names)\n",
    "coords_df.to_csv(\"/disk2/user/hilpek/project_root/scRNAseq/input2/BeadLocationsForR.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "daae4197",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# load barcodes name\n",
    "barcodes = adata_vis.obs_names\n",
    "\n",
    "# Convert coords to DataFrame, assign barcodes as index\n",
    "coords = pd.DataFrame(coords, index=barcodes, columns=[\"x\", \"y\"])\n",
    "\n",
    "# Convert indexes to R-compatible format (with dot and .1)\n",
    "coords.index = coords.index.to_series().str.replace(r\"^(V10F03)-\", r\"\\1.\", regex=True)\n",
    "coords.index = coords.index.to_series().str.replace(r\"-1$\", \".1\", regex=True)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cell2_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.23"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
